### PR TITLE
Migrate SM session metrics to `mongoose_instrument`

### PR DESCRIPTION
**Pull request by jacekwegr**
This PR includes the following:

- Changes the way we track user sessions by switching from `mongoose_metrics_probe` to `mongoose_instrument_probe`
- Updates the tests to support these changed metrics
- Updates the documentation about changed metrics 
- The metric `[global, cache, unique_sessions_number]` was removed because it relied on a specific metric backend and wasn't used anywhere else in the code

### CODE REVIEW

**Comment by mongoose-im** [comment]
[elasticsearch_and_cassandra_26](https://circleci.com/gh/esl/MongooseIM/225228) / elasticsearch_and_cassandra_mnesia / 724ab63e17b7038d2f15b0e336f7b645f993e919
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4313/225228/elasticsearch_and_cassandra_mnesia.26.2.2-amd64)/ [big](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4313/225228/elasticsearch_and_cassandra_mnesia.26.2.2-amd64/big/ct_run.test%409f6b873080b3.2024-06-27_17.05.33/index.html)
**OK: 457** / Failed: 0 / **User-skipped: 41** / Auto-skipped: 0

**Comment by mongoose-im** [comment]
[elasticsearch_and_cassandra_26](https://circleci.com/gh/esl/MongooseIM/225334) / elasticsearch_and_cassandra_mnesia / 4a003e18addf656234bd96b924f9742862cfa829
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4313/225334/elasticsearch_and_cassandra_mnesia.26.2.2-amd64)/ [big](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4313/225334/elasticsearch_and_cassandra_mnesia.26.2.2-amd64/big/ct_run.test%407ca34f5a43cc.2024-06-28_16.26.14/index.html)
**OK: 457** / Failed: 0 / **User-skipped: 41** / Auto-skipped: 0

**Comment by mongoose-im** [comment]
[elasticsearch_and_cassandra_26](https://circleci.com/gh/esl/MongooseIM/225360) / elasticsearch_and_cassandra_mnesia / b58cdc54dbc6ae4b7825b8a44d6074a6b186b2a8
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4313/225360/elasticsearch_and_cassandra_mnesia.26.2.2-amd64)/ [big](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4313/225360/elasticsearch_and_cassandra_mnesia.26.2.2-amd64/big/ct_run.test%4028ac434d1185.2024-06-28_16.49.24/index.html)
**OK: 456** / **Failed: 1** / **User-skipped: 41** / Auto-skipped: 0 
<details><summary>mam_SUITE:cassandra_muc_all:muc_prefs_cases:muc_prefs_set_cdata_request</summary>


```erlang
{error,
  {function_clause,
    [{mam_helper,parse_children_prefs_iq,
       [{xmlel,<<"error">>,
          [{<<"code">>,<<"501">>},{<<"type">>,<<"cancel">>}],
          [{xmlel,<<"feature-not-implemented">>,
             [{<<"xmlns">>,
             <<"urn:ietf:params:xml:ns:xmpp-stanzas">>}],
             []},
           {xmlel,<<"text">>,
             [{<<"xmlns">>,
             <<"urn:ietf:params:xml:ns:xmpp-stanzas">>}],
             [{xmlcdata,<<"From mod_mam_muc">>}]}]},
        {prefs_result_iq,undefined,[],[]}],
       []},
     {lists,foldl,3,[{file,"lists.erl"},{line,1594}]},
     {mam_SUITE,'-muc_prefs_set_cdata_request/1-fun-0-',2,
       [{file,"/home/circleci/project/big_tests/tests/mam_SUITE.erl"},
        {line,3457}]},
     {escalus_story,story,4,
       [{file,
          "/home/circleci/project/big_tests/_build/default/lib/escalus/src/escalus_story.erl"},
        {line,72}]},
     {muc_helper,story_with_room,4,
       [{file,"/home/circleci/project/big_tests/tests/muc_helper.erl"},
        {line,318}]},
     {test_server,ts_tc,3,[{file,"test_server.erl"},{line,1793}]},
     {test_server,run_test_case_eval1,6,
       [{file,"test_server.erl"},{line,1302}]},
     {test_server,run_test_case_eval,9,
       [{file,"test_server.erl"},{line,1234}]}]}}
```

[Report log](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4313/225360/elasticsearch_and_cassandra_mnesia.26.2.2-amd64/big/ct_run.test%4028ac434d1185.2024-06-28_16.49.24/big_tests.tests.mam_SUITE.logs/run.2024-06-28_16.49.25/mam_suite.muc_prefs_set_cdata_request.6532.html)
</details>

**Comment by mongoose-im** [comment]
[elasticsearch_and_cassandra_26](https://circleci.com/gh/esl/MongooseIM/225401) / elasticsearch_and_cassandra_mnesia / c9d896b52c27ab4cf9c07dbb18f7340344911d97
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4313/225401/elasticsearch_and_cassandra_mnesia.26.2.2-amd64)/ [big](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4313/225401/elasticsearch_and_cassandra_mnesia.26.2.2-amd64/big/ct_run.test%40fae4298baaea.2024-06-28_17.00.20/index.html)
**OK: 456** / **Failed: 1** / **User-skipped: 41** / Auto-skipped: 0 
<details><summary>mam_SUITE:elasticsearch_muc_all:muc_prefs_cases:muc_prefs_set_cdata_request</summary>


```erlang
{error,
  {function_clause,
    [{mam_helper,parse_children_prefs_iq,
       [{xmlel,<<"error">>,
          [{<<"code">>,<<"501">>},{<<"type">>,<<"cancel">>}],
          [{xmlel,<<"feature-not-implemented">>,
             [{<<"xmlns">>,
             <<"urn:ietf:params:xml:ns:xmpp-stanzas">>}],
             []},
           {xmlel,<<"text">>,
             [{<<"xmlns">>,
             <<"urn:ietf:params:xml:ns:xmpp-stanzas">>}],
             [{xmlcdata,<<"From mod_mam_muc">>}]}]},
        {prefs_result_iq,undefined,[],[]}],
       []},
     {lists,foldl,3,[{file,"lists.erl"},{line,1594}]},
     {mam_SUITE,'-muc_prefs_set_cdata_request/1-fun-0-',2,
       [{file,"/home/circleci/project/big_tests/tests/mam_SUITE.erl"},
        {line,3457}]},
     {escalus_story,story,4,
       [{file,
          "/home/circleci/project/big_tests/_build/default/lib/escalus/src/escalus_story.erl"},
        {line,72}]},
     {muc_helper,story_with_room,4,
       [{file,"/home/circleci/project/big_tests/tests/muc_helper.erl"},
        {line,318}]},
     {test_server,ts_tc,3,[{file,"test_server.erl"},{line,1793}]},
     {test_server,run_test_case_eval1,6,
       [{file,"test_server.erl"},{line,1302}]},
     {test_server,run_test_case_eval,9,
       [{file,"test_server.erl"},{line,1234}]}]}}
```

[Report log](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4313/225401/elasticsearch_and_cassandra_mnesia.26.2.2-amd64/big/ct_run.test%40fae4298baaea.2024-06-28_17.00.20/big_tests.tests.mam_SUITE.logs/run.2024-06-28_17.00.21/mam_suite.muc_prefs_set_cdata_request.19459.html)
</details>

File: src/ejabberd_sm.erl [hunk 1]
```diff
@@ -312,16 +315,16 @@ get_session_pid(JID) ->
 
 -spec get_unique_sessions_number() -> integer().
 get_unique_sessions_number() ->
+    InternalDB = get_internal_db(),
     try
         C = ejabberd_sm_backend:unique_count(),
-        mongoose_metrics:update(global, ?UNIQUE_COUNT_CACHE, C),
+        set_cached_unique_count(C, InternalDB),
```
Comment (chrzaszcz) [inline src/ejabberd_sm.erl:321]: At the grooming we agreed that we could just remove this cache entirely, what's the reason for the change? Is it used anywhere?

File: src/metrics/mongoose_metrics_probe_node_sessions.erl [hunk 2]
```diff
@@ -14,12 +14,11 @@
 %% limitations under the License.
 %%==============================================================================
 -module(mongoose_metrics_probe_node_sessions).
--behaviour(mongoose_metrics_probe).
+-behaviour(mongoose_instrument_probe).
```
Comment (chrzaszcz) [inline src/metrics/mongoose_metrics_probe_node_sessions.erl:17]: We need to rename and move all three probe modules, because the goal is to empty the whole `mongoose_metrics` directory. We also don't want to give an impression, that this is still the old probe.

File: src/ejabberd_sm.erl [hunk 3]
```diff
@@ -484,14 +487,19 @@ init([]) ->
                   end,
                   ?ALL_HOST_TYPES),
     %% Create metrics after backend has started, otherwise probe could have null value
-    create_metrics(),
+    start_probes(),
```
Comment (chrzaszcz) [inline src/ejabberd_sm.erl:490]: I think such names are short-lived, because you are assuming that all global metrics are probes, and an addition of a new metric would change that. I'd rather change line 488 to have `[global | ?ALL_HOST_TYPES]`.

**Review by chrzaszcz — COMMENTED** [review]
It's going in the right direction. I added a few comments.

**Comment by mongoose-im** [comment]
[elasticsearch_and_cassandra_26](https://circleci.com/gh/esl/MongooseIM/225958) / elasticsearch_and_cassandra_mnesia / bf2e07bf6c39424db536e110741d214fa17928f3
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4313/225958/elasticsearch_and_cassandra_mnesia.26.2.2-amd64)/ [big](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4313/225958/elasticsearch_and_cassandra_mnesia.26.2.2-amd64/big/ct_run.test%40feaab39e775b.2024-07-02_14.12.25/index.html)
**OK: 451** / **Failed: 6** / **User-skipped: 41** / Auto-skipped: 0 
<details><summary>mam_SUITE:cassandra_prefs_cases:messages_filtered_when_prefs_default_policy_is_roster</summary>


```erlang
{error,{test_case_failed,"ASSERT EQUAL\n\tExpected []\n\tValue [ok,ok,ok]\n"}}
```

[Report log](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4313/225958/elasticsearch_and_cassandra_mnesia.26.2.2-amd64/big/ct_run.test%40feaab39e775b.2024-07-02_14.12.25/big_tests.tests.mam_SUITE.logs/run.2024-07-02_14.12.26/mam_suite.messages_filtered_when_prefs_default_policy_is_roster.6658.html)
</details>
<details><summary>mam_SUITE:elasticsearch_prefs_cases:messages_filtered_when_prefs_default_policy_is_roster</summary>


```erlang
{error,{test_case_failed,"ASSERT EQUAL\n\tExpected []\n\tValue [ok,ok,ok]\n"}}
```

[Report log](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4313/225958/elasticsearch_and_cassandra_mnesia.26.2.2-amd64/big/ct_run.test%40feaab39e775b.2024-07-02_14.12.25/big_tests.tests.mam_SUITE.logs/run.2024-07-02_14.12.26/mam_suite.messages_filtered_when_prefs_default_policy_is_roster.14340.html)
</details>
<details><summary>gdpr_SUITE:retrieve_personal_data:retrieve_roster</summary>


```erlang
{error,{{assertion_failed,assert_many,false,[is_presence],[],[]},
    [{escalus_new_assert,assert_true,2,
               [{file,"/home/circleci/project/big_tests/_build/default/lib/escalus/src/escalus_new_assert.erl"},
                {line,84}]},
     {escalus_story,'-make_all_clients_friends/1-fun-0-',2,
            [{file,"/home/circleci/project/big_tests/_build/default/lib/escalus/src/escalus_story.erl"},
             {line,109}]},
     {escalus_utils,'-each_with_index/3-fun-0-',3,
            [{file,"/home/circleci/project/big_tests/_build/default/lib/escalus/src/escalus_utils.erl"},
             {line,87}]},
     {lists,foldl_1,3,[{file,"lists.erl"},{line,1599}]},
     {escalus_utils,'-each_with_index/3-fun-0-',3,
            [{file,"/home/circleci/project/big_tests/_build/default/lib/escalus/src/escalus_utils.erl"},
             {line,87}]},
     {lists,foldl,3,[{file,"lists.erl"},{line,1594}]},
     {escalus_utils,distinct_pairs,2,
            [{file,"/home/circleci/project/big_tests/_build/default/lib/escalus/src/escalus_utils.erl"},
             {line,60}]},
     {escalus_story,make_all_clients_friends,1,
            [{file,"/home/circleci/project/big_tests/_build/default/lib/escalus/src/escalus_story.erl"},
             {line,106}]}]}}
```

[Report log](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4313/225958/elasticsearch_and_cassandra_mnesia.26.2.2-amd64/big/ct_run.test%40feaab39e775b.2024-07-02_14.12.25/big_tests.tests.gdpr_SUITE.logs/run.2024-07-02_14.15.24/gdpr_suite.retrieve_roster.html)
</details>
<details><summary>gdpr_SUITE:retrieve_personal_data:retrieve_offline</summary>


```erlang
{error,{{timeout,3,[{times,50,0}],ok},
    [{mongoose_helper,do_wait_until,2,
              [{file,"/home/circleci/project/big_tests/../test/common/mongoose_helper.erl"},
               {line,359}]},
     {gdpr_SUITE,'-retrieve_offline/1-fun-3-',4,
           [{file,"/home/circleci/project/big_tests/tests/gdpr_SUITE.erl"},
            {line,950}]},
     {escalus_story,story,4,
            [{file,"/home/circleci/project/big_tests/_build/default/lib/escalus/src/escalus_story.erl"},
             {line,72}]},
     {test_server,ts_tc,3,[{file,"test_server.erl"},{line,1793}]},
     {test_server,run_test_case_eval1,6,
            [{file,"test_server.erl"},{line,1302}]},
     {test_server,run_test_case_eval,9,
            [{file,"test_server.erl"},{line,1234}]}]}}
```

[Report log](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4313/225958/elasticsearch_and_cassandra_mnesia.26.2.2-amd64/big/ct_run.test%40feaab39e775b.2024-07-02_14.12.25/big_tests.tests.gdpr_SUITE.logs/run.2024-07-02_14.15.24/gdpr_suite.retrieve_offline.html)
</details>
<details><summary>gdpr_SUITE:remove_personal_data:remove_roster</summary>


```erlang
{error,{{assertion_failed,assert_many,false,[is_presence],[],[]},
    [{escalus_new_assert,assert_true,2,
               [{file,"/home/circleci/project/big_tests/_build/default/lib/escalus/src/escalus_new_assert.erl"},
                {line,84}]},
     {escalus_story,'-make_all_clients_friends/1-fun-0-',2,
            [{file,"/home/circleci/project/big_tests/_build/default/lib/escalus/src/escalus_story.erl"},
             {line,109}]},
     {escalus_utils,'-each_with_index/3-fun-0-',3,
            [{file,"/home/circleci/project/big_tests/_build/default/lib/escalus/src/escalus_utils.erl"},
             {line,87}]},
     {lists,foldl_1,3,[{file,"lists.erl"},{line,1599}]},
     {escalus_utils,'-each_with_index/3-fun-0-',3,
            [{file,"/home/circleci/project/big_tests/_build/default/lib/escalus/src/escalus_utils.erl"},
             {line,87}]},
     {lists,foldl,3,[{file,"lists.erl"},{line,1594}]},
     {escalus_utils,distinct_pairs,2,
            [{file,"/home/circleci/project/big_tests/_build/default/lib/escalus/src/escalus_utils.erl"},
             {line,60}]},
     {escalus_story,make_all_clients_friends,1,
            [{file,"/home/circleci/project/big_tests/_build/default/lib/escalus/src/escalus_story.erl"},
             {line,106}]}]}}
```

[Report log](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4313/225958/elasticsearch_and_cassandra_mnesia.26.2.2-amd64/big/ct_run.test%40feaab39e775b.2024-07-02_14.12.25/big_tests.tests.gdpr_SUITE.logs/run.2024-07-02_14.15.24/gdpr_suite.remove_roster.html)
</details>
<details><summary>gdpr_SUITE:remove_personal_data:remove_offline</summary>


```erlang
{error,{{timeout,3,[{times,50,0}],ok},
    [{mongoose_helper,do_wait_until,2,
              [{file,"/home/circleci/project/big_tests/../test/common/mongoose_helper.erl"},
               {line,359}]},
     {gdpr_SUITE,'-remove_offline/1-fun-1-',4,
           [{file,"/home/circleci/project/big_tests/tests/gdpr_SUITE.erl"},
            {line,985}]},
     {escalus_story,story,4,
            [{file,"/home/circleci/project/big_tests/_build/default/lib/escalus/src/escalus_story.erl"},
             {line,72}]},
     {test_server,ts_tc,3,[{file,"test_server.erl"},{line,1793}]},
     {test_server,run_test_case_eval1,6,
            [{file,"test_server.erl"},{line,1302}]},
     {test_server,run_test_case_eval,9,
            [{file,"test_server.erl"},{line,1234}]}]}}
```

[Report log](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4313/225958/elasticsearch_and_cassandra_mnesia.26.2.2-amd64/big/ct_run.test%40feaab39e775b.2024-07-02_14.12.25/big_tests.tests.gdpr_SUITE.logs/run.2024-07-02_14.15.24/gdpr_suite.remove_offline.html)
</details>

**Comment by mongoose-im** [comment]
[elasticsearch_and_cassandra_26](https://circleci.com/gh/esl/MongooseIM/225990) / elasticsearch_and_cassandra_mnesia / 12e84b1b6b7b53b02cde4290d79dc518862377b3
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4313/225990/elasticsearch_and_cassandra_mnesia.26.2.2-amd64)/ [big](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4313/225990/elasticsearch_and_cassandra_mnesia.26.2.2-amd64/big/ct_run.test%40960283bca166.2024-07-02_16.21.24/index.html)
**OK: 454** / **Failed: 1** / **User-skipped: 41** / Auto-skipped: 0 
<details><summary>mam_SUITE:elasticsearch_muc_all:muc_prefs_cases:muc_messages_filtered_when_prefs_default_policy_is_never</summary>


```erlang
{error,{test_case_failed,"ASSERT EQUAL\n\tExpected []\n\tValue [ok]\n"}}
```

[Report log](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4313/225990/elasticsearch_and_cassandra_mnesia.26.2.2-amd64/big/ct_run.test%40960283bca166.2024-07-02_16.21.24/big_tests.tests.mam_SUITE.logs/run.2024-07-02_16.21.26/mam_suite.muc_messages_filtered_when_prefs_default_policy_is_never.24194.html)
</details>

**Comment by mongoose-im** [comment]
[elasticsearch_and_cassandra_26](https://circleci.com/gh/esl/MongooseIM/226130) / elasticsearch_and_cassandra_mnesia / 146aa7226ccbe4ebf88d2a4319e45d736214d09b
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4313/226130/elasticsearch_and_cassandra_mnesia.26.2.2-amd64)/ [big](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4313/226130/elasticsearch_and_cassandra_mnesia.26.2.2-amd64/big/ct_run.test%406e0e88358e82.2024-07-03_08.38.59/index.html)
**OK: 455** / Failed: 0 / **User-skipped: 41** / Auto-skipped: 0

**Review by chrzaszcz — APPROVED** [review]
Looks good 👌
